tests/sentry/spans/test_buffer_v2.py

```diff
@@ -0,0 +1,188 @@
+    wiped by the test harness.
+    """
+    assert not client.keys("*")
+
+
+def test_basic(buffer: RedisSpansBufferV2):
+    spans = [
+        Span(
+            payload=b"A", trace_id="a" * 32, span_id="c" * 16, parent_span_id="b" * 16, project_id=1
+        ),
+        Span(
+            payload=b"B", trace_id="a" * 32, span_id="d" * 16, parent_span_id="b" * 16, project_id=1
+        ),
+        Span(
+            payload=b"C", trace_id="a" * 32, span_id="e" * 16, parent_span_id="b" * 16, project_id=1
+        ),
+        Span(payload=b"D", trace_id="a" * 32, span_id="b" * 16, parent_span_id=None, project_id=1),
+    ]
+
+    buffer.process_spans(spans, now=0)
+
+    assert_ttls(buffer.client)
+
+    assert buffer.flush_segments(now=5) == {}
+    rv = buffer.flush_segments(now=11)
+    assert rv == {buffer._segment_id(1, "a" * 32, "b" * 16): {b"D", b"B", b"A", b"C"}}
+    assert buffer.flush_segments(now=30) == {}
+
+    buffer.done_flush_segments(rv)
+
+    assert_clean(buffer.client)
+
+
+def test_parent_first(buffer: RedisSpansBufferV2):
+    spans = [
+        Span(payload=b"D", trace_id="a" * 32, span_id="b" * 16, parent_span_id=None, project_id=1),
+        Span(
+            payload=b"C", trace_id="a" * 32, span_id="e" * 16, parent_span_id="b" * 16, project_id=1
+        ),
+        Span(
+            payload=b"B", trace_id="a" * 32, span_id="d" * 16, parent_span_id="b" * 16, project_id=1
+        ),
+        Span(
+            payload=b"A", trace_id="a" * 32, span_id="c" * 16, parent_span_id="b" * 16, project_id=1
+        ),
+    ]
+
+    buffer.process_spans(spans, now=0)
+
+    assert_ttls(buffer.client)
+
+    assert buffer.flush_segments(now=10) == {}
+    assert buffer.flush_segments(now=20) == {}
+    rv = buffer.flush_segments(now=60)
+    assert rv == {buffer._segment_id(1, "a" * 32, "b" * 16): {b"D", b"B", b"A", b"C"}}
+
+    buffer.done_flush_segments(rv)
+
+    assert_clean(buffer.client)
+
+
+def test_parent_middle(buffer: RedisSpansBufferV2):
+    spans = [
+        Span(
+            payload=b"C", trace_id="a" * 32, span_id="e" * 16, parent_span_id="b" * 16, project_id=1
+        ),
+        Span(
+            payload=b"B", trace_id="a" * 32, span_id="d" * 16, parent_span_id="b" * 16, project_id=1
+        ),
+        Span(payload=b"D", trace_id="a" * 32, span_id="b" * 16, parent_span_id=None, project_id=1),
+        Span(
+            payload=b"A", trace_id="a" * 32, span_id="c" * 16, parent_span_id="b" * 16, project_id=1
+        ),
+    ]
+
+    buffer.process_spans(spans, now=0)
+
+    assert_ttls(buffer.client)
+
+    assert buffer.flush_segments(now=10) == {}
+    assert buffer.flush_segments(now=20) == {}
+    rv = buffer.flush_segments(now=60)
+    assert rv == {buffer._segment_id(1, "a" * 32, "b" * 16): {b"D", b"B", b"A", b"C"}}
+
+    buffer.done_flush_segments(rv)
+
+    assert_clean(buffer.client)
+
+
+def test_parent_in_other_project(buffer: RedisSpansBufferV2):
```

Comment:
How does the expected behavior change when a trace includes multiple projects?

Reply:
the behavior that is expected is that a segment is terminated when a parent span is crossing project boundaries:

```
span A (project=1)
  span B (project=2)
    span B1 (project=2)
  span C (project=2)
```

expected behavior is that span B, B1 and C go into the same segment, and A gets its own segment. at the moment the code creates segments `A`, `B+B1` and `C`.

https://www.notion.so/sentry/Spans-Pipeline-with-Dataworkers-18b8b10e4b5d8025a7adfadbfc3e9a19

specifically: https://www.notion.so/sentry/Spans-Pipeline-with-Dataworkers-18b8b10e4b5d8025a7adfadbfc3e9a19?pvs=4#1998b10e4b5d80bc8f86fe959efe9017